tests/unit/test_charm.py

```diff
@@ -907,3 +909,86 @@ def test_given_magma_service_enabled_when_install_then_nothing_done(
                 ),
             ]
         )
+
+    @patch("charm.Path")
+    @patch("subprocess.run")
+    def test_when_orchestrator_available_event_then_configuration_is_installed(
+        self, patch_subprocess_run, patch_path
+    ):
+        event = Mock(
+            root_ca_certificate="root_ca_certificate_content",
+            orchestrator_address="orchestrator_address",
+            orchestrator_port="orchestrator_port",
+            bootstrapper_address="bootstrapper_address",
+            bootstrapper_port="bootstrapper_port",
+            fluentd_address="fluentd_address",
+            fluentd_port="fluentd_port",
+        )
+        self.harness.charm._on_orchestrator_available(event)
+
+        patch_path.assert_has_calls(
+            [
+                call("/var/opt/magma/tmp/certs/rootCA.pem"),
+                call().write_text("root_ca_certificate_content"),
+                call("/var/opt/magma/configs/control_proxy.yml"),
+                call().write_text(
+                    "cloud_address: orchestrator_address\n"
+                    "cloud_port: orchestrator_port\n"
+                    "bootstrap_address: bootstrapper_address\n"
+                    "bootstrap_port: bootstrapper_port\n"
+                    "fluentd_address: fluentd_address\n"
+                    "fluentd_port: fluentd_port\n"
+                    "\n"
+                    "rootca_cert: /var/opt/magma/tmp/certs/rootCA.pem\n"
+                ),
+            ],
+            any_order=True,
+        )
+        patch_subprocess_run.assert_has_calls(
+            [
+                call(
+                    ["service", "magma@*", "stop"],
+                    stdout=-1,
+                ),
+                call(
+                    ["service", "magma@magmad", "start"],
+                    stdout=-1,
+                ),
+            ]
+        )
+
+    def test_given_directory_does_not_exist_when_install_file_then_directory_is_created(self):
+        with tempfile.TemporaryDirectory() as directory:
+            file = pathlib.Path(directory) / "does_not_exist" / "file.txt"
+
+            self.harness.charm._install_file(file, "content")
```

Comment:
As we discussed with Guillaume some time ago, we should avoid calling private methods/properties directly. From that perspective, I think creation of the `orchestrator` relation should be the starting point for this test.
What do you think?

Reply:
I agree with the general rule, but in this case, I think there is enough logic to test in that one unit to test it separately. I currently have 4 tests for that one method, and if I test it through the `orchestrator` relation, the number of tests will only go up, as will the complexity of those tests.

I could also move this method outside of the charm class and test it as a public utility function. WDYT?

Reply:
I'm not sure if having public utils is "charmingly" correct ;)
I also buy your arguments for keeping these tests the way they are. Let's forget it.

Reply:
I think I agree with moving the utility function outside of the class since it contains no Magma/AGW/charm specific logic.

Reply:
Fixed